_posts/2019/2019-07-28-consider-value-objects.md

```diff
@@ -0,0 +1,78 @@
+---
+title: "Consider Value Objects"
+description: "Making text readable on"
+layout: article
+category: ruby
+image:
+  base: '2019/consider-value-objects'
+  alt: 'Cubes'
+  credit: Christian Fregnan
+  source: "https://unsplash.com/photos/TAVKURx-xLw"
+
+---
+
+You’ll often find yourself doing creating multiple view helpers around the same concept, creating complex calculation methods, or passing a value (or several) around multiple methods.
+
+When this happens it is possible a simple object, smaller than an Active Record model, is trying to reveal itself.
+
+In these cases, consider refactoring the functionality into a “Value Object”.
+
+## Instead of...
+
+...just using a helper:
+
+```ruby
+def css_color(r, g, b)
+  "##{r.to_s(16)}#{g.to_s(16)}#{b.to_s(16)}"
+end
+```
+
+## Use...
+
+...an extracted “Value Object” to truly capture the concept your application is using.
+
+```ruby
+class RGBColor
+  def initialize(red, green, blue)
+    @red = [[0, red].max], 255].min
+    @green = [[0, green].max], 255].min
+    @blue = [[0, blue].max], 255].min      
+  end
+
+  def to_hex
+    [@red, @green, @blue].each_with_object("") do |part, to_hex|
+      to_hex << part.to_s(16)
+    end
+  end
+end
+```
+
+### And in the helper
+
+```ruby
+def css_color(r, g, b)
+  "##{RGBColor.new(r, g, b).to_hex}"
+end
+```
+
+
+## Why?
+
+Just because a concept in our app isn't saved to the database doesn't mean it shouldn’t be an object in our app.
+
+Martin Fowler [defines a Value Object](https://martinfowler.com/eaaCatalog/valueObject.html) as:
+
+> A small simple object, like money or a date range, whose equality isn’t based on identity.
+
+Refactoring into a value object has several benefits as you isolate the code for this concept. You might hear this referred to as ‘separation of concerns’.
+
+Your can test the behaviour of your concept in isolation, leading to more comprehensive and efficient tests.
+
+This improved organisation of your code will clarify your understanding. The code is also easier to reuse and you have a clear place for extending the functionality of the object.
+
+
+## Why not?
+
+Finding the correct time to extract a concept into a value object can be difficult. Too early leads to unnecessary complication and potentially increased confusion; too late and you’re already coding in a mess!
+
```

Comment:
Should this part be in the 'Why'? Since it's showing a way to implement it despite saying avoid `Struct` and `OpenStruct`?

I'm torn!

Reply:
What if it opens with "It might be too soon. Finding..."

Reply:
I was actually talking about the sentence afterwards, but this addition works really well.